File: plugins/state/inmemory-go/harness_test.go
// harness_test.go — conformance/golden-data harness for the state/v1 axis.
//
// Loads contracts/conformance/state-v1.json (the SAME file the Python reference
// loads) and drives StateService through the stub core-mediated gateway: Get/Put/
// List via unary Invoke, Watch via the ADR-008 InvokeServerStream relay. The
// lifecycle is STATEFUL; the errors array exercises the KEY GRAMMAR. Bad keys are
// built from key_len ("a"*N, oversize) / key_inject ("a"+chr(N)+"b", NUL/control)
// so the vector file stays pure-ASCII valid JSON.
package main

import (
	"context"
	"encoding/json"
	"io"
	"net"
	"os"
	"strings"
	"testing"
	"time"

	commonv1 "github.com/rat-dev/rat/gen/rat/common/v1"
	corev1 "github.com/rat-dev/rat/gen/rat/core/v1"
	statev1 "github.com/rat-dev/rat/gen/rat/state/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/grpc/test/bufconn"
	"google.golang.org/protobuf/proto"
)

// ---- golden-vector model (mirrors contracts/conformance/state-v1.json) ----

type watchEventExpect struct {
	Type     string `json:"type"`
	Key      string `json:"key"`
	Revision int64  `json:"revision"`
}

type expectation struct {
	Found       *bool              `json:"found"`
	Value       string             `json:"value"`
	Revision    *int64             `json:"revision"`
	Outcome     string             `json:"outcome"`
	Keys        []string           `json:"keys"`
	WatchEvents []watchEventExpect `json:"watch_events"`
	Code        string             `json:"code"`
}

type vstep struct {
	Step         string      `json:"step"`
	Op           string      `json:"op"`
	Key          string      `json:"key"`
	Value        string      `json:"value"`
	Prefix       string      `json:"prefix"`
	IfRevision   int64       `json:"if_revision"`
	FromRevision int64       `json:"from_revision"`
	KeyLen       int         `json:"key_len"`
	KeyInject    *int        `json:"key_inject"`
	Expect       expectation `json:"expect"`
}

type vectors struct {
	Axis      string  `json:"axis"`
	Lifecycle []vstep `json:"lifecycle"`
	Errors    []vstep `json:"errors"`
}

const vectorPath = "../../../contracts/conformance/state-v1.json"

func loadVectors(t *testing.T) vectors {
	t.Helper()
	raw, err := os.ReadFile(vectorPath)
	if err != nil {
		t.Fatalf("read golden vectors %s: %v", vectorPath, err)
	}
	var v vectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("parse golden vectors: %v", err)
	}
	if v.Axis != "state/v1" {
		t.Fatalf("vectors axis = %q, want state/v1", v.Axis)
	}
	return v
}

// resolveKey builds the request key: key_len → "a"*N (oversize); key_inject →
// "a"+chr(N)+"b" (NUL/control); else the literal key.
func resolveKey(s vstep) string {
	if s.KeyLen > 0 {
		return strings.Repeat("a", s.KeyLen)
	}
	if s.KeyInject != nil {
		return "a" + string(rune(*s.KeyInject)) + "b"
	}
	return s.Key
}

func outcomeEnum(t *testing.T, name string) statev1.PutOutcome {
	switch name {
	case "COMMITTED":
		return statev1.PutOutcome_PUT_OUTCOME_COMMITTED
	case "CONFLICT":
		return statev1.PutOutcome_PUT_OUTCOME_CONFLICT
	case "UNKNOWN":
		return statev1.PutOutcome_PUT_OUTCOME_UNKNOWN
	default:
		t.Fatalf("unmapped outcome %q", name)
		return statev1.PutOutcome_PUT_OUTCOME_UNSPECIFIED
	}
}

// ---- harness wiring: plugin behind the stub core gateway ----

type rig struct {
	gw   corev1.CapabilityInvokeServiceClient
	core *stubGateway
}

func bufDial(t *testing.T, lis *bufconn.Listener) *grpc.ClientConn {
	t.Helper()
	conn, err := grpc.NewClient(
		"passthrough:///bufnet",
		grpc.WithContextDialer(func(_ context.Context, _ string) (net.Conn, error) { return lis.Dial() }),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		t.Fatalf("dial: %v", err)
	}
	t.Cleanup(func() { _ = conn.Close() })
	return conn
}

func newRig(t *testing.T) *rig {
	t.Helper()

	plis := bufconn.Listen(1 << 20)
	psrv := grpc.NewServer()
	statev1.RegisterStateServiceServer(psrv, newServer())
	go func() { _ = psrv.Serve(plis) }()
	t.Cleanup(psrv.Stop)
	providerConn := bufDial(t, plis)

	core := newGateway(providerConn, "rat-strategy-test", []string{
		"rat://state/v1/get", "rat://state/v1/put", "rat://state/v1/list", "rat://state/v1/watch",
		"rat://state/v1/create-if-absent",
	})

	glis := bufconn.Listen(1 << 20)
	gsrv := grpc.NewServer()
	corev1.RegisterCapabilityInvokeServiceServer(gsrv, core)
	go func() { _ = gsrv.Serve(glis) }()
	t.Cleanup(gsrv.Stop)

	return &rig{gw: corev1.NewCapabilityInvokeServiceClient(bufDial(t, glis)), core: core}
}

func tctx(t *testing.T) context.Context {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	t.Cleanup(cancel)
	return ctx
}

func withCallMeta(ctx context.Context) context.Context {
	rc := &commonv1.RequestContext{
		Trace:    &commonv1.TraceContext{Traceparent: "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01", CorrelationId: "corr-golden"},
		Identity: &commonv1.Identity{Tenant: "acme"},
	}
	b, _ := proto.Marshal(rc)
	return metadata.AppendToOutgoingContext(ctx, callMetaHeader, string(b))
}

func (r *rig) invoke(ctx context.Context, capURI string, req, resp proto.Message) error {
	payload, err := proto.Marshal(req)
	if err != nil {
		return err
	}
	out, err := r.gw.Invoke(withCallMeta(ctx), &corev1.InvokeRequest{Capability: capURI, Payload: payload})
	if err != nil {
		return err
	}
	return proto.Unmarshal(out.GetResult(), resp)
}

func (r *rig) watch(ctx context.Context, prefix string, fromRev int64) ([]*statev1.WatchResponse, error) {
	payload, err := proto.Marshal(&statev1.WatchRequest{Prefix: prefix, FromRevision: fromRev})
	if err != nil {
		return nil, err
	}
	stream, err := r.gw.InvokeServerStream(withCallMeta(ctx), &corev1.InvokeServerStreamRequest{
		Capability: "rat://state/v1/watch", Payload: payload,
	})
	if err != nil {
		return nil, err
	}
	var events []*statev1.WatchResponse
	for {
		frame, err := stream.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			return events, err
		}
		var ev statev1.WatchResponse
		if err := proto.Unmarshal(frame.GetResult(), &ev); err != nil {
			return events, err
		}
		events = append(events, &ev)
	}
	return events, nil
}

func runStep(t *testing.T, r *rig, ctx context.Context, s vstep) {
	t.Helper()
	switch s.Op {
	case "get":
		var resp statev1.GetResponse
		err := r.invoke(ctx, "rat://state/v1/get", &statev1.GetRequest{Key: resolveKey(s)}, &resp)
		if expectedError(t, s, err) {
			return
		}
		assertGet(t, s, &resp)
	case "put":
		var resp statev1.PutResponse
		err := r.invoke(ctx, "rat://state/v1/put",
			&statev1.PutRequest{Key: resolveKey(s), Value: []byte(s.Value), IfRevision: s.IfRevision}, &resp)
		if expectedError(t, s, err) {
			return
		}
		assertPut(t, s, &resp)
	case "list":
		var resp statev1.ListResponse
		err := r.invoke(ctx, "rat://state/v1/list", &statev1.ListRequest{Prefix: s.Prefix}, &resp)
		if expectedError(t, s, err) {
			return
		}
		assertKeys(t, s, resp.GetKeys())
	case "watch":
		events, err := r.watch(ctx, s.Prefix, s.FromRevision)
		if expectedError(t, s, err) {
			return
		}
		assertWatch(t, s, events)
	case "create-if-absent":
		var resp statev1.CreateIfAbsentResponse
		err := r.invoke(ctx, "rat://state/v1/create-if-absent",
			&statev1.CreateIfAbsentRequest{Key: resolveKey(s), Value: []byte(s.Value)}, &resp)
		if expectedError(t, s, err) {
			return
		}
		e := s.Expect
		if e.Outcome != "" && resp.GetOutcome() != outcomeEnum(t, e.Outcome) {
			t.Fatalf("%s: outcome = %s, want %s", s.Step, resp.GetOutcome(), e.Outcome)
		}
		if e.Revision != nil && resp.GetRevision() != *e.Revision {
			t.Fatalf("%s: revision = %d, want %d", s.Step, resp.GetRevision(), *e.Revision)
		}
	default:
		t.Fatalf("%s: unknown op %q", s.Step, s.Op)
	}
}

// ---- the tests ----

func TestStateConformance_GoldenVectors(t *testing.T) {
	v := loadVectors(t)
	r := newRig(t)
	ctx := tctx(t)

	for _, s := range v.Lifecycle {
		s := s
		t.Run(s.Step, func(t *testing.T) { runStep(t, r, ctx, s) })
	}

	if got, want := len(r.core.auditLog()), len(v.Lifecycle); got != want {
		t.Fatalf("audit log = %d entries, want one per mediated call (%d)", got, want)
	}
}

func TestStateConformance_ErrorVectors(t *testing.T) {
	v := loadVectors(t)
	r := newRig(t)
	ctx := tctx(t)

	for _, s := range v.Errors {
		s := s
		t.Run(s.Step, func(t *testing.T) { runStep(t, r, ctx, s) })
	}
}

// ---- assertions ----

func expectedError(t *testing.T, s vstep, err error) bool {
	t.Helper()
	if s.Expect.Code != "" {
		if err == nil {
			t.Fatalf("%s: want error %s, got nil", s.Step, s.Expect.Code)
		}
		if got := status.Code(err); got != wantCode(t, s.Expect.Code) {
			t.Fatalf("%s: status = %s, want %s", s.Step, got, s.Expect.Code)
		}
		return true
	}
	if err != nil {
		t.Fatalf("%s: unexpected error: %v", s.Step, err)
	}
	return false
}

func assertGet(t *testing.T, s vstep, resp *statev1.GetResponse) {
	t.Helper()
	e := s.Expect
	if e.Found != nil && resp.GetFound() != *e.Found {
		t.Fatalf("%s: found = %v, want %v", s.Step, resp.GetFound(), *e.Found)
	}
	if e.Value != "" && string(resp.GetValue()) != e.Value {
		t.Fatalf("%s: value = %q, want %q", s.Step, string(resp.GetValue()), e.Value)
	}
	if e.Revision != nil && resp.GetRevision() != *e.Revision {
		t.Fatalf("%s: revision = %d, want %d", s.Step, resp.GetRevision(), *e.Revision)
	}
}

func assertPut(t *testing.T, s vstep, resp *statev1.PutResponse) {
	t.Helper()
	e := s.Expect
	if e.Outcome != "" && resp.GetOutcome() != outcomeEnum(t, e.Outcome) {
		t.Fatalf("%s: outcome = %s, want %s", s.Step, resp.GetOutcome(), e.Outcome)
	}
	if e.Revision != nil && resp.GetRevision() != *e.Revision {
		t.Fatalf("%s: revision = %d, want %d", s.Step, resp.GetRevision(), *e.Revision)
	}
}

func assertKeys(t *testing.T, s vstep, got []string) {
	t.Helper()
	want := s.Expect.Keys
	if len(got) != len(want) {
		t.Fatalf("%s: keys = %v, want %v", s.Step, got, want)
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("%s: keys = %v, want %v", s.Step, got, want)
		}
	}
}

func assertWatch(t *testing.T, s vstep, events []*statev1.WatchResponse) {
	t.Helper()
	want := s.Expect.WatchEvents
	if len(events) != len(want) {
		t.Fatalf("%s: watch events = %d, want %d", s.Step, len(events), len(want))
	}
	for i, w := range want {
		ev := events[i]
		if ev.GetType() != watchType(t, w.Type) || ev.GetKey() != w.Key || ev.GetRevision() != w.Revision {
			t.Fatalf("%s: event[%d] = {%s %s rev %d}, want {%s %s rev %d}",
				s.Step, i, ev.GetType(), ev.GetKey(), ev.GetRevision(), w.Type, w.Key, w.Revision)
		}
	}
}

func watchType(t *testing.T, name string) statev1.WatchResponse_Type {
	t.Helper()
	switch name {
	case "PUT":
		return statev1.WatchResponse_TYPE_PUT
	case "DELETE":
		return statev1.WatchResponse_TYPE_DELETE
	default:
		t.Fatalf("unmapped watch type %q", name)
		return statev1.WatchResponse_TYPE_UNSPECIFIED
	}
}

func wantCode(t *testing.T, name string) codes.Code {
	t.Helper()
	switch name {
	case "INVALID_ARGUMENT":
		return codes.InvalidArgument
	case "NOT_FOUND":
		return codes.NotFound
	case "FAILED_PRECONDITION":
		return codes.FailedPrecondition
	case "PERMISSION_DENIED":
		return codes.PermissionDenied
	default:
		t.Fatalf("unmapped expected code %q", name)
		return codes.Unknown
	}
}

File: plugins/state/inmemory-py/harness_test.py
"""Conformance/golden-data harness for the state/v1 axis — Python side.

The ADR-003 cross-run: loads the SAME golden vectors the Go reference loads
(contracts/conformance/state-v1.json) and drives THIS independent implementation's
StateService over real gRPC (Get/Put/List + a streaming Watch). The lifecycle is
STATEFUL; the errors array exercises the KEY GRAMMAR. Bad keys are built from
key_len / key_inject so the vector file stays pure-ASCII. Driven directly (no
gateway), like the other Python references. Runs standalone or under pytest.
"""

import json
import os
from concurrent import futures

import grpc

from rat.state.v1 import state_pb2, state_pb2_grpc

from server import StateServicer

VECTOR_PATH = os.path.join(
    os.path.dirname(__file__), "..", "..", "..", "contracts", "conformance", "state-v1.json"
)

_CODE = {
    "INVALID_ARGUMENT": grpc.StatusCode.INVALID_ARGUMENT,
    "NOT_FOUND": grpc.StatusCode.NOT_FOUND,
    "FAILED_PRECONDITION": grpc.StatusCode.FAILED_PRECONDITION,
    "PERMISSION_DENIED": grpc.StatusCode.PERMISSION_DENIED,
}

_OUTCOME = {
    "COMMITTED": state_pb2.PutOutcome.PUT_OUTCOME_COMMITTED,
    "CONFLICT": state_pb2.PutOutcome.PUT_OUTCOME_CONFLICT,
    "UNKNOWN": state_pb2.PutOutcome.PUT_OUTCOME_UNKNOWN,
}

_WTYPE = {
    "PUT": state_pb2.WatchResponse.TYPE_PUT,
    "DELETE": state_pb2.WatchResponse.TYPE_DELETE,
}


def load_vectors():
    with open(VECTOR_PATH, "r", encoding="utf-8") as f:
        v = json.load(f)
    assert v["axis"] == "state/v1", f'vectors axis = {v["axis"]!r}, want state/v1'
    return v


def _resolve_key(s):
    if "key_len" in s:
        return "a" * s["key_len"]
    if "key_inject" in s:
        return "a" + chr(s["key_inject"]) + "b"
    return s.get("key", "")


class Rig:
    def __init__(self) -> None:
        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=4))
        state_pb2_grpc.add_StateServiceServicer_to_server(StateServicer(), self.server)
        port = self.server.add_insecure_port("127.0.0.1:0")
        self.server.start()
        self.channel = grpc.insecure_channel(f"127.0.0.1:{port}")
        self.stub = state_pb2_grpc.StateServiceStub(self.channel)

    def close(self) -> None:
        self.channel.close()
        self.server.stop(None)

    def do(self, s):
        op = s["op"]
        if op == "get":
            return self.stub.Get(state_pb2.GetRequest(key=_resolve_key(s)))
        if op == "put":
            return self.stub.Put(state_pb2.PutRequest(
                key=_resolve_key(s), value=s.get("value", "").encode("utf-8"),
                if_revision=s.get("if_revision", 0)))
        if op == "list":
            return self.stub.List(state_pb2.ListRequest(prefix=s.get("prefix", "")))
        if op == "watch":
            return list(self.stub.Watch(
                state_pb2.WatchRequest(prefix=s.get("prefix", ""), from_revision=s.get("from_revision", 0))))
        if op == "create-if-absent":
            return self.stub.CreateIfAbsent(state_pb2.CreateIfAbsentRequest(
                key=_resolve_key(s), value=s.get("value", "").encode("utf-8")))
        raise AssertionError(f'unknown op {op!r}')


def _assert_success(s, resp):
    e, op = s["expect"], s["op"]
    if op == "get":
        if "found" in e:
            assert resp.found == e["found"], f'found = {resp.found}, want {e["found"]}'
        if "value" in e:
            assert resp.value.decode("utf-8") == e["value"], f'value = {resp.value!r}, want {e["value"]!r}'
        if "revision" in e:
            assert resp.revision == e["revision"], f'revision = {resp.revision}, want {e["revision"]}'
    elif op in ("put", "create-if-absent"):
        if "outcome" in e:
            assert resp.outcome == _OUTCOME[e["outcome"]], f'outcome = {resp.outcome}, want {e["outcome"]}'
        if "revision" in e:
            assert resp.revision == e["revision"], f'revision = {resp.revision}, want {e["revision"]}'
    elif op == "list":
        assert list(resp.keys) == e["keys"], f'keys = {list(resp.keys)}, want {e["keys"]}'
    elif op == "watch":
        want = e["watch_events"]
        assert len(resp) == len(want), f'watch events = {len(resp)}, want {len(want)}'
        for ev, w in zip(resp, want):
            assert ev.type == _WTYPE[w["type"]] and ev.key == w["key"] and ev.revision == w["revision"], (
                f'event = {{{ev.type} {ev.key} rev {ev.revision}}}, want {w}')


def run_step(rig: Rig, s):
    code = s["expect"].get("code")
    try:
        resp = rig.do(s)
    except grpc.RpcError as err:
        assert code, f'{s["step"]}: unexpected error {err.code()}'
        assert err.code() == _CODE[code], f'{s["step"]}: status {err.code()}, want {code}'
        return
    assert not code, f'{s["step"]}: want error {code}, got success'
    _assert_success(s, resp)


def run_lifecycle(rig: Rig, v) -> None:
    for s in v["lifecycle"]:
        run_step(rig, s)


def run_errors(rig: Rig, v) -> None:
    for s in v["errors"]:
        run_step(rig, s)


def test_golden_vectors():
    rig = Rig()
    try:
        run_lifecycle(rig, load_vectors())
    finally:
        rig.close()


def test_error_vectors():
    rig = Rig()
    try:
        run_errors(rig, load_vectors())
    finally:
        rig.close()


if __name__ == "__main__":
    v = load_vectors()
    rig = Rig()
    try:
        run_lifecycle(rig, v)
    finally:
        rig.close()
    rig = Rig()
    try:
        run_errors(rig, v)
    finally:
        rig.close()
    print("PASS — rat-state-inmemory-py conformed to state/v1 golden vectors")
